src/protocol/prf_sharding/mod.rs

```diff
@@ -0,0 +1,704 @@
+        }
+        i += 1;
+    }
+
+    Ok(output)
+}
+
+fn initialize_new_device_attribution_variables<BK, TV>(
+    share_of_one: Replicated<Gf2>,
+    input_row: &PrfShardedIpaInputRow<BK, TV>,
+    num_breakdown_key_bits: usize,
+    num_trigger_value_bits: usize,
+    num_saturating_sum_bits: usize,
+) -> InputsRequiredFromPrevRow
+where
+    BK: GaloisField,
+    TV: GaloisField,
+{
+    InputsRequiredFromPrevRow {
+        ever_encountered_a_source_event: share_of_one - &input_row.is_trigger_bit,
+        attributed_breakdown_key_bits: BitDecomposed::decompose(num_breakdown_key_bits, |i| {
+            Replicated::new(
+                Gf2::truncate_from(input_row.breakdown_key.left()[i]),
+                Gf2::truncate_from(input_row.breakdown_key.right()[i]),
+            )
+        }),
+        saturating_sum: BitDecomposed::new(vec![Replicated::ZERO; num_saturating_sum_bits]),
+        is_saturated: Replicated::ZERO,
+        // This is incorrect in the case that the CAP is less than the maximum value of "trigger value" for a single row
+        // Not a problem if you assume that's an invalid input
+        difference_to_cap: BitDecomposed::new(vec![Replicated::ZERO; num_trigger_value_bits]),
+    }
+}
+
+///
+/// Returns (`sum_bit`, `carry_out`)
+///
+async fn one_bit_adder<C, SB>(
+    ctx: C,
+    record_id: RecordId,
+    x: &SB,
+    y: &SB,
+    carry_in: &SB,
+) -> Result<(SB, SB), Error>
+where
+    C: UpgradedContext<Gf2, Share = SB>,
+    SB: LinearSecretSharing<Gf2> + BasicProtocols<C, Gf2>,
+{
+    // compute sum bit as x XOR y XOR carry_in
+    let sum_bit = x.clone() + y + carry_in;
+
+    let x_xor_carry_in = x.clone() + carry_in;
+    let y_xor_carry_in = y.clone() + carry_in;
+    let carry_out = x_xor_carry_in
+        .multiply(&y_xor_carry_in, ctx, record_id)
+        .await?
+        + carry_in;
+
+    Ok((sum_bit, carry_out))
+}
+
+async fn compute_saturating_sum<C, SB>(
+    ctx: C,
+    record_id: RecordId,
+    cur_value: &BitDecomposed<SB>,
+    prev_sum: &BitDecomposed<SB>,
+    prev_is_saturated: &SB,
+    num_trigger_value_bits: usize,
+    num_saturating_sum_bits: usize,
+) -> Result<(BitDecomposed<SB>, SB), Error>
+where
+    C: UpgradedContext<Gf2, Share = SB>,
+    SB: LinearSecretSharing<Gf2> + BasicProtocols<C, Gf2>,
+{
+    assert!(cur_value.len() == num_trigger_value_bits);
+    assert!(prev_sum.len() == num_saturating_sum_bits);
+
+    let mut carry_in = SB::ZERO;
+    let mut output = vec![];
+    for i in 0..num_saturating_sum_bits {
+        let c = ctx.narrow(&BitOpStep::from(i));
+        let (sum_bit, carry_out) = if i < num_trigger_value_bits {
+            one_bit_adder(c, record_id, &cur_value[i], &prev_sum[i], &carry_in).await?
+        } else {
+            one_bit_adder(c, record_id, &SB::ZERO, &prev_sum[i], &carry_in).await?
+        };
+
+        output.push(sum_bit);
+        carry_in = carry_out;
+    }
+    let updated_is_saturated = -carry_in
+        .clone()
```

Comment:
I concur with @akoshelev's suggestion to encapsulate the saturating sum. I might suggest using `protocol::boolean::or` instead of the inline logic, but I think instead it's possible to compute `is_saturated` and `just_saturated` using only one multiply:

```
just_saturated = carry & !prev_saturated
is_saturated = just_saturated + prev_saturated
```

(Where the `+` on the second line is addition, not logical OR, because they can't both be true.)

Reply:
Woah! Nice! That's a slick optimization that I will definitely put in. Thanks for the careful review Andy!

Reply:
OK, now that I have encapsulated this in a SaturatingSum class, I no longer have access to `prev_saturated` and I'm not sure how to put this optimization in.

I suggest leaving it out for now, and saving it for a rainy day when we need a little boost in our performance =).